src/components/views/Instance/Edit/InstancePermissions.tsx

```diff
@@ -0,0 +1,862 @@
+                this.setState(
+                    prevState =>
+                        (({
+                            [permTargetField]: {
+                                ...(prevState[permTargetField] as Record<string, Permission>),
+                                [key]: {
+                                    currentVal: currentVal,
+                                    bitflag: val
+                                }
+                            }
+                        } as unknown) as IState)
+                );
+            });
+        };
+
+        loadEnum(
+            InstancePermissionSetRights,
+            "instancePermissionSetRights",
+            "permsinstancepermissionset"
+        );
+        loadEnum(RepositoryRights, "repositoryRights", "permsrepository");
+        loadEnum(ByondRights, "byondRights", "permsbyond");
+        loadEnum(DreamMakerRights, "dreamMakerRights", "permsdreammaker");
+        loadEnum(DreamDaemonRights, "dreamDaemonRights", "permsdreamdaemon");
+        loadEnum(ChatBotRights, "chatBotRights", "permschatbots");
+        loadEnum(ConfigurationRights, "configurationRights", "permsconfiguration");
+    }
+
+    private addError(error: InternalError<ErrorCode>): void {
+        this.setState(prevState => {
+            const errors = Array.from(prevState.errors);
+            errors.push(error);
+            return {
+                errors
+            };
+        });
+    }
+
+    private async loadCurrentPermissions(selectedPermissionSetId: number) {
+        this.setState({
+            loadingPerms: true
+        });
+
+        if (selectedPermissionSetId === resolvePermissionSet(this.context.user).id) {
+            // This can hit the cache and seem "too" fast.
+            // slow it down a little
+            const sanicSpeedLimit = new Promise(r => setTimeout(r, 200));
+            const response = await InstancePermissionSetClient.getCurrentInstancePermissionSet(
+                this.context.instance.id,
+                true
+            );
+            await sanicSpeedLimit;
+            if (response.code === StatusCode.OK) {
+                this.setState({
+                    userPermissions: response.payload
+                });
+
+                this.loadEnums(response.payload);
+            } else {
+                this.addError(response.error);
+            }
+        } else if (
+            hasInstancePermRight(
+                this.context.instancePermissionSet,
+                InstancePermissionSetRights.Read
+            )
+        ) {
+            const response = await InstancePermissionSetClient.getByPermissionSetId(
+                this.context.instance.id,
+                selectedPermissionSetId
+            );
+            if (response.code === StatusCode.OK) {
+                this.loadEnums(response.payload);
+            } else if (response.error.code != ErrorCode.NO_DB_ENTITY) {
```

Comment:
Why would this be a not equals instead of being a default case?

Reply:
So we can handle the 410 (missing instancepermissionset) explicitly? There's an else here, no?